File: backend_samples/library/javascript/AzureAIVisionFaceDeviceAttestation/src/api_telemetry.ts
/**
 * Framework-agnostic per-route failure telemetry.
 *
 * `trackApiFail(route, reason, status, props?)` emits a structured customEvent
 * so failure reasons can be grouped/queried by route + reason code. Handlers
 * call it directly; it has no HTTP-framework dependency.
 *
 * The Next.js request-telemetry wrapper (`withApiTelemetry`) lives app-side in
 * app/_lib/next_telemetry.ts.
 */
import { trackEvent } from './logging';

export function trackApiFail(
  route: string,
  reason: string,
  status: number,
  properties?: Record<string, unknown>
): void {
  trackEvent('Api.Fail', {
    route,
    reason,
    status,
    ...properties,
  });
}

File: backend_samples/library/python/AzureAIVisionFaceDeviceAttestation/azure_ai_vision_face_deviceattestation/api_telemetry.py
"""Framework-agnostic per-route failure telemetry.

Only the structured failure event lives in the library. The FastAPI
request-wrapping decorator (``with_api_telemetry``) is host glue and lives in the
sample, not here.
"""

from __future__ import annotations

from typing import Any, Optional

from .logging import track_event


def track_api_fail(
    route: str,
    reason: str,
    status: int,
    properties: Optional[dict[str, Any]] = None,
) -> None:
    props: dict[str, Any] = {"route": route, "reason": reason, "status": status}
    if properties:
        props.update(properties)
    track_event("Api.Fail", props)
